pkg/virt-controller/watch/vm_initializer.go

```diff
@@ -0,0 +1,371 @@
+	"kubevirt.io/kubevirt/pkg/log"
+)
+
+type VirtualMachineInitializer struct {
+	vmPresetInformer cache.SharedIndexInformer
+	clientset        kubecli.KubevirtClient
+	queue            workqueue.RateLimitingInterface
+	store            cache.Store
+}
+
+const initializerMarking = "presets.virtualmachines.kubevirt.io"
+
+func NewVirtualMachineInitializer(vmPresetInformer cache.SharedIndexInformer, queue workqueue.RateLimitingInterface, vmInitCache cache.Store, clientset kubecli.KubevirtClient) *VirtualMachineInitializer {
+	vmi := VirtualMachineInitializer{
+		vmPresetInformer: vmPresetInformer,
+		clientset:        clientset,
+		queue:            queue,
+		store:            vmInitCache,
+	}
+	return &vmi
+}
+
+func (c *VirtualMachineInitializer) Run(threadiness int, stopCh chan struct{}) {
+	defer controller.HandlePanic()
+	defer c.queue.ShutDown()
+	log.Log.Info("Starting Virtual Machine Initializer.")
+
+	// Wait for cache sync before we start the pod controller
+	cache.WaitForCacheSync(stopCh, c.vmPresetInformer.HasSynced)
+
+	// Start the actual work
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	log.Log.Info("Stopping controller.")
+}
+
+func (c *VirtualMachineInitializer) runWorker() {
+	for c.Execute() {
+	}
+}
+
+func (c *VirtualMachineInitializer) Execute() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	err := c.execute(key.(string))
+
+	if err != nil {
+		log.Log.Reason(err).Infof("reenqueuing VM %v", key)
+		c.queue.AddRateLimited(key)
+	} else {
+		log.Log.V(4).Infof("processed VM %v", key)
+		c.queue.Forget(key)
+	}
+	return true
+}
+
+func (c *VirtualMachineInitializer) execute(key string) error {
+
+	// Fetch the latest Vm state from cache
+	obj, exists, err := c.store.GetByKey(key)
+
+	if err != nil {
+		return err
+	}
+
+	// Retrieve the VM
+	var vm *kubev1.VirtualMachine
+	if !exists {
+		namespace, name, err := cache.SplitMetaNamespaceKey(key)
+		if err != nil {
+			return err
+		}
+		vm = kubev1.NewVMReferenceFromNameWithNS(namespace, name)
+	} else {
+		vm = obj.(*kubev1.VirtualMachine)
+	}
+
+	// only process VM's that aren't initialized by this controller yet
+	if !isInitialized(vm) {
+		return c.initializeVirtualMachine(vm)
+	}
+	// Ignore fully initialized Virtual Machines
+	return nil
+}
+
+func (c *VirtualMachineInitializer) initializeVirtualMachine(vm *kubev1.VirtualMachine) error {
+	// All VM's must be marked as initialized or they are held in limbo forever
+	// Collect all errors and defer returning until after the update
+	logger := log.Log
+	errors := []error{}
+
+	logger.Object(vm).Info("Initializing VirtualMachine")
+
+	allPresets, err := listPresets(c.vmPresetInformer, vm.GetNamespace())
+
+	matchingPresets, err := filterPresets(allPresets, vm)
+
+	if err != nil {
+		logger.Object(vm).Reason(err).Errorf("Error while matching presets to VirtualMachine")
+		errors = append(errors, err)
```

Comment:
What's the point of proceeding with processing the presets if filterPresets failed?

Reply:
As an initializer, this controller is responsible for marking every VM as initialized. The controller cannot bail out early for any error condition because it is responsible for removing the initializer marking. Without that step, the VM simply disappears--the client hangs and things get confusing.

Reply:
My understanding is Initialization takes place after object persistence.  We should be able to return an error and rate limit the key in order to try again. What would make the VM disappear if we return an error?